src/lib/analytics/segmentOneTrustIntegration.ts

```diff
@@ -0,0 +1,258 @@
+    "Segment.io": consentArray.some(d => d === SEGMENT_MAPPING),
+  }
+
+  // for each destination, if its id maps to a OneTrust category that is present in consent, set it true, to be enabled.
+  const additionalPref = destinations
+    .map(function (dest) {
+      if (
+        dest.id in segmentToOneTrust &&
+        consentArray.includes(segmentToOneTrust[dest.id])
+      ) {
+        return { [dest.id]: true }
+      } else {
+        // no mapping for the destination, or it maps to a OneTrust category that is not present in OneTrust consent.
+        return { [dest.id]: false }
+      }
+    })
+    .reduce((acc, val) => {
+      return {
+        ...val,
+        ...acc,
+      }
+    }, [])
+
+  const destinationPref = {
+    ...basePref,
+    ...additionalPref,
+  }
+
+  return destinationPref
+}
+
+function conditionallyLoadAnalytics({
+  writeKey,
+  destinationPref,
+  shouldReload = true,
+}) {
+  let isAnythingEnabled = false
+
+  for (const destination in destinationPref) {
+    const isEnabled = destinationPref[destination]
+    if (isEnabled) {
+      isAnythingEnabled = true
+    }
+  }
+
+  // @ts-ignore
+  if (window.analytics.initialized) {
+    // Segment has been initialized, a page reload is required to reload Segment.
+    if (shouldReload) {
+      window.location.reload()
+    }
+    return
+  }
+
+  if (isAnythingEnabled) {
+    // @ts-ignore
+    window.analytics.load(writeKey, { integrations: destinationPref })
+  }
+
+  // if everything is disabled, don't load segment.
+}
+
+async function fetchDestinationForWriteKey(writeKey) {
+  try {
+    const res = await window.fetch(
+      `https://cdn.segment.com/v1/projects/${writeKey}/integrations`
+    )
+
+    if (!res.ok) {
+      throw new Error(`HTTP ${res.status} ${res.statusText}`)
+    }
+
+    const destinations = await res.json()
+
+    if (!validateSegmentResponse(destinations)) {
+      throw new Error("Segment integrations API endpoint returned invalid data")
+    }
+
+    // Rename creationName to id to abstract the weird data model
+    for (const destination of destinations) {
+      destination.id = destination.creationName
+      delete destination.creationName
+    }
+
+    return destinations
+  } catch (error) {
+    console.error("Error: fetching Segment destinations resulted in: ", error)
+    return []
+  }
+}
+
+function validateSegmentResponse(destinations) {
+  if (!Array.isArray(destinations)) {
+    return false
+  }
+
+  for (const destination of destinations) {
+    if (!("creationName" in destination)) {
+      return false
+    }
+
+    if (typeof destination.creationName !== "string") {
+      return false
+    }
+  }
+
+  return true
+}
+
+export const exportForTesting = {
+  conditionallyLoadAnalytics,
+  delay,
+  fetchDestinationForWriteKey,
+  fetchDestinations,
+  getConsent,
+  getConsentAndLoad,
+  getOneTrustConsent,
+  oneTrustReady,
+  setSegmentDestinationPref,
+  validateSegmentResponse,
+}
```

Comment:
just want to 👍 this pattern! I use it a lot and haven't seen others do it before now. Generally people make everything public through exports even though said exports aren't actually used. By scoping things in this way it keeps things nice and focused.